modules/shortcodes/instagram.php

```diff
@@ -133,27 +133,111 @@ function jetpack_instagram_embed_reversal( $content ) {
+ * @param string $url  URL of the content to be embedded.
+ * @param array  $atts Shortcode attributes.
  *
- * @return string
+ * @return array $params Array of parameters to be used in Instagram query.
  */
+function jetpack_instagram_get_allowed_parameters( $url, $atts = array() ) {
+	global $content_width;
+
+	// Any URL passed via a shortcode attribute takes precedence.
+	if ( ! empty( $atts['url'] ) ) {
+		$url = $atts['url'];
+		unset( $atts['url'] );
+	}
+
+	/*
+	 * Get URL and parameters from the URL if possible.
+	 *
+	 * We'll also clean any other query params from the URL since Facebook's new API for Instagram
+	 * embeds does not like query parameters. See p7H4VZ-2DU-p2.
+	 */
+	$parsed_url = wp_parse_url( $url );
+	if ( $parsed_url && isset( $parsed_url['host'] ) && isset( $parsed_url['path'] ) ) {
+		// Bail early if this is not an Instagram URL.
+		if ( ! preg_match( '/(?:^|\.)instagr(?:\.am|am\.com)$/', $parsed_url['host'] ) ) {
+			return array();
+		}
+
+		$url = 'https://www.instagram.com' . $parsed_url['path'];
+
+		// If we have any parameters as part of the URL, we merge them with our attributes.
+		if ( ! empty( $parsed_url['query'] ) ) {
+			$query_args = array();
+			wp_parse_str( $parsed_url['query'], $query_args );
+
+			$atts = array_merge( $atts, $query_args );
+		}
+	} else {
+		return array();
+	}
+
+	$max_width = 698;
+	$min_width = 320;
+
+	$params = shortcode_atts(
+		array(
+			'url'         => $url,
+			'width'       => isset( $content_width ) ? $content_width : $max_width,
+			'height'      => '',
+			'hidecaption' => false,
+		),
+		$atts,
+		'instagram'
+	);
+
+	// Ensure width is within bounds.
+	$params['width'] = absint( $params['width'] );
+	if ( $params['width'] > $max_width ) {
+		$params['width'] = $max_width;
+	} elseif ( $params['width'] < $min_width ) {
+		$params['width'] = $min_width;
+	}
+
+	return $params;
+}
+
+	/**
+	 * Add auth token required by Instagram's oEmbed REST API, or proxy through WP.com.
+	 *
+	 * @since 9.1.0
+	 *
+	 * @param string $provider URL of the oEmbed provider.
+	 * @param string $url      URL of the content to be embedded.
+	 *
+	 * @return string
+	 */
 function jetpack_instagram_oembed_fetch_url( $provider, $url ) {
 	if ( ! wp_startswith( $provider, 'https://graph.facebook.com/v5.0/instagram_oembed/' ) ) {
 		return $provider;
 	}
 
-	// Attempt to clean query params from the URL since Facebook's new API for Instagram
-	// embeds does not like query parameters. See p7H4VZ-2DU-p2.
-	$parsed_url = wp_parse_url( $url );
-	if ( $parsed_url ) {
-		$clean_url            = 'https://www.instagram.com' . $parsed_url['path'];
+	// Get a set of URL and parameters supported by Facebook.
+	$clean_parameters = jetpack_instagram_get_allowed_parameters( $url );
+
+	if ( ! empty( $clean_parameters['url'] ) ) {
+		// Replace existing URL by our clean version.
 		$provider_without_url = remove_query_arg( 'url', $provider );
```

Comment:
Why do you do `remove_query_arg()` here, but not below for the other args?

Some quick testing suggests that it's not needed here, `add_query_arg()` seems to replace any existing arg with the same name.

Reply:
Yeah, it does seem okay without it. Removed in 06ed85ae1deed0775f2b7b42113972db1d90b75c